File: examples/ahc015/python/game.py
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

SIDE = 10
CELL_COUNT = SIDE * SIDE
ACTION_COUNT = 4

FRONT = 0
BACK = 1
LEFT = 2
RIGHT = 3

Board = NDArray[np.uint8]


def empty_board() -> Board:
    return np.zeros((SIDE, SIDE), dtype=np.uint8)


def place_at_rank(board: Board, rank: int, flavor: int) -> Board:
    empty = np.flatnonzero(board.reshape(-1) == 0)
    if rank < 1 or rank > len(empty):
        raise ValueError(f"rank {rank} is outside 1..={len(empty)}")
    result = board.copy()
    result.reshape(-1)[empty[rank - 1]] = flavor
    return result


def place_at_ranks(
    boards: NDArray[np.uint8],
    ranks: NDArray[np.integer],
    flavors: NDArray[np.integer],
) -> NDArray[np.uint8]:
    """Place one item in every board without a Python loop over episodes."""
    boards_array = np.asarray(boards, dtype=np.uint8)
    ranks_array = np.asarray(ranks).reshape(-1)
    flavors_array = np.asarray(flavors, dtype=np.uint8).reshape(-1)
    if boards_array.ndim != 3 or boards_array.shape[1:] != (SIDE, SIDE):
        raise ValueError("boards must have shape [episodes, 10, 10]")
    if len(boards_array) != len(ranks_array) or len(boards_array) != len(flavors_array):
        raise ValueError("batched placement fields have different lengths")

    flat = boards_array.reshape(len(boards_array), CELL_COUNT)
    empty = flat == 0
    valid = (ranks_array >= 1) & (ranks_array <= empty.sum(axis=1))
    if not np.all(valid):
        raise ValueError("rank is outside the remaining empty cells")
    selected = np.argmax(np.cumsum(empty, axis=1) == ranks_array[:, None], axis=1)
    result = flat.copy()
    result[np.arange(len(result)), selected] = flavors_array
    return result.reshape(-1, SIDE, SIDE)


def inserted(board: Board, cell: int, flavor: int) -> Board:
    if cell < 0 or cell >= CELL_COUNT or board.reshape(-1)[cell] != 0:
        raise ValueError(f"cell {cell} is not empty")
    result = board.copy()
    result.reshape(-1)[cell] = flavor
    return result


def tilt(board: Board, action: int) -> Board:
    result = empty_board()
    if action in (FRONT, BACK):
        for column in range(SIDE):
            values = board[:, column]
            values = values[values != 0]
            if action == FRONT:
                result[: len(values), column] = values
            else:
                result[SIDE - len(values) :, column] = values
    elif action in (LEFT, RIGHT):
        for row in range(SIDE):
            values = board[row, :]
            values = values[values != 0]
            if action == LEFT:
                result[row, : len(values)] = values
            else:
                result[row, SIDE - len(values) :] = values
    else:
        raise ValueError(f"invalid action: {action}")
    return result


def afterstates(board: Board) -> NDArray[np.uint8]:
    return np.stack([tilt(board, action) for action in range(ACTION_COUNT)])


def tilt_batch(boards: NDArray[np.uint8], action: int) -> NDArray[np.uint8]:
    """Apply one stable tilt to a batch of boards using NumPy indexing."""
    boards_array = np.asarray(boards, dtype=np.uint8)
    if boards_array.ndim != 3 or boards_array.shape[1:] != (SIDE, SIDE):
        raise ValueError("boards must have shape [episodes, 10, 10]")
    mask = boards_array != 0
    episode_indices, row_indices, column_indices = np.nonzero(mask)
    result = np.zeros_like(boards_array)
    if action == FRONT:
        destinations = np.cumsum(mask, axis=1) - 1
        destination_rows = destinations[episode_indices, row_indices, column_indices]
        result[episode_indices, destination_rows, column_indices] = boards_array[mask]
    elif action == BACK:
        counts = mask.sum(axis=1)
        destinations = SIDE - counts[:, None, :] + np.cumsum(mask, axis=1) - 1
        destination_rows = destinations[episode_indices, row_indices, column_indices]
        result[episode_indices, destination_rows, column_indices] = boards_array[mask]
    elif action == LEFT:
        destinations = np.cumsum(mask, axis=2) - 1
        destination_columns = destinations[episode_indices, row_indices, column_indices]
        result[episode_indices, row_indices, destination_columns] = boards_array[mask]
    elif action == RIGHT:
        counts = mask.sum(axis=2)
        destinations = SIDE - counts[:, :, None] + np.cumsum(mask, axis=2) - 1
        destination_columns = destinations[episode_indices, row_indices, column_indices]
        result[episode_indices, row_indices, destination_columns] = boards_array[mask]
    else:
        raise ValueError(f"invalid action: {action}")
    return result


def afterstates_batch(boards: NDArray[np.uint8]) -> NDArray[np.uint8]:
    return np.stack([tilt_batch(boards, action) for action in range(ACTION_COUNT)], axis=1)


def connectivity_numerator(board: Board) -> int:
    visited = np.zeros((SIDE, SIDE), dtype=np.bool_)
    numerator = 0
    for start_row in range(SIDE):
        for start_column in range(SIDE):
            flavor = int(board[start_row, start_column])
            if flavor == 0 or visited[start_row, start_column]:
                continue
            visited[start_row, start_column] = True
            stack = [(start_row, start_column)]
            size = 0
            while stack:
                row, column = stack.pop()
                size += 1
                for next_row, next_column in (
                    (row - 1, column),
                    (row + 1, column),
                    (row, column - 1),
                    (row, column + 1),
                ):
                    if (
                        0 <= next_row < SIDE
                        and 0 <= next_column < SIDE
                        and not visited[next_row, next_column]
                        and board[next_row, next_column] == flavor
                    ):
                        visited[next_row, next_column] = True
                        stack.append((next_row, next_column))
            numerator += size * size
    return numerator


def flavor_totals(flavors: NDArray[np.uint8]) -> NDArray[np.int64]:
    if flavors.shape != (CELL_COUNT,) or np.any((flavors < 1) | (flavors > 3)):
        raise ValueError("flavors must contain 100 values in 1..=3")
    return np.bincount(flavors, minlength=4)[1:].astype(np.int64)


def denominator(flavors: NDArray[np.uint8]) -> int:
    totals = flavor_totals(flavors)
    return int(np.dot(totals, totals))


def potential(board: Board, score_denominator: int) -> float:
    return connectivity_numerator(board) / score_denominator


def official_score(board: Board, score_denominator: int) -> int:
    return int(np.floor(1_000_000 * potential(board, score_denominator) + 0.5))


@dataclass(slots=True)
class EpisodeState:
    flavors: NDArray[np.uint8]
    board: Board
    placed: int = 0

    @classmethod
    def new(cls, flavors: NDArray[np.uint8]) -> EpisodeState:
        flavor_totals(flavors)
        return cls(flavors.copy(), empty_board())

    @property
    def denominator(self) -> int:
        return denominator(self.flavors)

    def place_cell(self, cell: int) -> None:
        self.board = inserted(self.board, cell, int(self.flavors[self.placed]))
        self.placed += 1

    def place_rank(self, rank: int) -> None:
        self.board = place_at_rank(self.board, rank, int(self.flavors[self.placed]))
        self.placed += 1

    def apply(self, action: int) -> None:
        self.board = tilt(self.board, action)

File: examples/ahc015/rust/src/game.rs
pub const SIDE: usize = 10;
pub const CANDY_COUNT: usize = SIDE * SIDE;
pub const ACTION_COUNT: usize = 4;

pub type Board = [u8; CANDY_COUNT];
pub type Action = usize;

pub const FRONT: Action = 0;
pub const BACK: Action = 1;
pub const LEFT: Action = 2;
pub const RIGHT: Action = 3;

pub fn action_char(action: Action) -> char {
    match action {
        FRONT => 'F',
        BACK => 'B',
        LEFT => 'L',
        RIGHT => 'R',
        _ => panic!("invalid action: {action}"),
    }
}

#[derive(Clone, Debug)]
pub struct Input {
    flavors: [u8; CANDY_COUNT],
    totals: [usize; 3],
    first_positions: [usize; 3],
}

impl Input {
    pub fn new(flavors: [u8; CANDY_COUNT]) -> Self {
        let mut totals = [0; 3];
        let mut first_positions = [usize::MAX; 3];
        for (index, &flavor) in flavors.iter().enumerate() {
            assert!((1..=3).contains(&flavor));
            let flavor = flavor as usize - 1;
            totals[flavor] += 1;
            first_positions[flavor] = first_positions[flavor].min(index);
        }
        Self {
            flavors,
            totals,
            first_positions,
        }
    }

    pub fn flavors(&self) -> &[u8; CANDY_COUNT] {
        &self.flavors
    }

    pub fn totals(&self) -> [usize; 3] {
        self.totals
    }

    pub fn first_positions(&self) -> [usize; 3] {
        self.first_positions
    }
}

#[derive(Clone, Debug)]
pub struct State {
    board: Board,
    placed: usize,
}

impl State {
    pub fn new() -> Self {
        Self {
            board: [0; CANDY_COUNT],
            placed: 0,
        }
    }

    pub fn placed(&self) -> usize {
        self.placed
    }

    pub fn is_terminal(&self) -> bool {
        self.placed == CANDY_COUNT
    }

    pub fn place_at_rank(&mut self, rank: usize, flavor: u8) {
        assert!((1..=CANDY_COUNT - self.placed).contains(&rank));
        let mut empty_rank = 0;
        for cell in 0..CANDY_COUNT {
            if self.board[cell] == 0 {
                empty_rank += 1;
                if empty_rank == rank {
                    self.board[cell] = flavor;
                    self.placed += 1;
                    return;
                }
            }
        }
        unreachable!("validated rank must identify an empty cell");
    }

    pub fn afterstates(&self) -> [Board; ACTION_COUNT] {
        std::array::from_fn(|action| tilt(&self.board, action))
    }

    pub fn apply_action(&mut self, action: Action) {
        self.board = tilt(&self.board, action);
    }
}

pub const fn cell(row: usize, column: usize) -> usize {
    row * SIDE + column
}

pub fn tilt(board: &Board, action: Action) -> Board {
    let mut result = [0; CANDY_COUNT];
    match action {
        FRONT => {
            for column in 0..SIDE {
                let mut destination = 0;
                for row in 0..SIDE {
                    let flavor = board[cell(row, column)];
                    if flavor != 0 {
                        result[cell(destination, column)] = flavor;
                        destination += 1;
                    }
                }
            }
        }
        BACK => {
            for column in 0..SIDE {
                let mut destination = SIDE;
                for row in (0..SIDE).rev() {
                    let flavor = board[cell(row, column)];
                    if flavor != 0 {
                        destination -= 1;
                        result[cell(destination, column)] = flavor;
                    }
                }
            }
        }
        LEFT => {
            for row in 0..SIDE {
                let mut destination = 0;
                for column in 0..SIDE {
                    let flavor = board[cell(row, column)];
                    if flavor != 0 {
                        result[cell(row, destination)] = flavor;
                        destination += 1;
                    }
                }
            }
        }
        RIGHT => {
            for row in 0..SIDE {
                let mut destination = SIDE;
                for column in (0..SIDE).rev() {
                    let flavor = board[cell(row, column)];
                    if flavor != 0 {
                        destination -= 1;
                        result[cell(row, destination)] = flavor;
                    }
                }
            }
        }
        _ => panic!("invalid action: {action}"),
    }
    result
}

pub fn place_on_board_at_rank(board: &Board, rank: usize, flavor: u8) -> Board {
    let mut result = *board;
    let mut empty_rank = 0;
    for cell in &mut result {
        if *cell == 0 {
            empty_rank += 1;
            if empty_rank == rank {
                *cell = flavor;
                return result;
            }
        }
    }
    panic!("rank {rank} does not identify an empty cell");
}

pub fn connectivity_numerator(board: &Board) -> usize {
    let mut visited = [false; CANDY_COUNT];
    let mut stack = [0usize; CANDY_COUNT];
    let mut numerator = 0;
    for start in 0..CANDY_COUNT {
        let flavor = board[start];
        if flavor == 0 || visited[start] {
            continue;
        }
        let mut stack_size = 1;
        stack[0] = start;
        visited[start] = true;
        let mut component_size = 0;
        while stack_size > 0 {
            stack_size -= 1;
            let current = stack[stack_size];
            component_size += 1;
            let row = current / SIDE;
            let column = current % SIDE;
            let mut neighbors = [usize::MAX; 4];
            if row > 0 {
                neighbors[0] = cell(row - 1, column);
            }
            if row + 1 < SIDE {
                neighbors[1] = cell(row + 1, column);
            }
            if column > 0 {
                neighbors[2] = cell(row, column - 1);
            }
            if column + 1 < SIDE {
                neighbors[3] = cell(row, column + 1);
            }
            for next in neighbors {
                if next != usize::MAX && !visited[next] && board[next] == flavor {
                    visited[next] = true;
                    stack[stack_size] = next;
                    stack_size += 1;
                }
            }
        }
        numerator += component_size * component_size;
    }
    numerator
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tilt_preserves_order_and_compacts() {
        let mut board = [0; CANDY_COUNT];
        board[cell(2, 0)] = 1;
        board[cell(7, 0)] = 2;
        board[cell(1, 3)] = 3;
        let front = tilt(&board, FRONT);
        assert_eq!(front[cell(0, 0)], 1);
        assert_eq!(front[cell(1, 0)], 2);
        assert_eq!(front[cell(0, 3)], 3);
        let right = tilt(&board, RIGHT);
        assert_eq!(right[cell(2, 9)], 1);
        assert_eq!(right[cell(7, 9)], 2);
    }

    #[test]
    fn rank_is_row_major_after_a_tilt() {
        let mut state = State::new();
        state.place_at_rank(1, 1);
        state.apply_action(RIGHT);
        state.place_at_rank(1, 2);
        assert_eq!(state.board[0], 2);
        assert_eq!(state.board[9], 1);
    }

    #[test]
    fn component_score_matches_hand_calculation() {
        let mut board = [0; CANDY_COUNT];
        board[cell(0, 0)] = 1;
        board[cell(0, 1)] = 1;
        board[cell(2, 2)] = 1;
        board[cell(4, 4)] = 2;
        board[cell(5, 4)] = 2;
        assert_eq!(connectivity_numerator(&board), 4 + 1 + 4);
    }
}
